### DISCUSSION TITLE
Apple M1 arm64 not exporting FMU

### DISCUSSION THREAD

**Discussion by arunraina2502** [OpenModelica/OpenModelica#10193] [Q&A] [answered]
Hi, 

On my Apple M1 Pro (macOS 13.1), I installed OpenModelica as per the instructions [here](https://github.com/OpenModelica/OpenModelica/blob/master/README.cmake.md#33-macos). I was able to create and simulate modelica classes and functions (tested with OMEdit). However, the exporting FMUs as per instructions [here](https://openmodelica.org/doc/OpenModelicaUsersGuide/1.20/fmitlm.html) doesn't work. From the very long log (lots of warnings), the errors I received are: 1. Error building simulator 2. -Bdynamic does not exist.

Any ideas?

**Comment by AnHeuermann** [comment, marked as answer]
Sounds like some configure is doing something wrong. Can you add the log file?

To pipe all of the omc output into one file you can create a file `test.mos`
```modelica
loadModel(Modelica); getErrorString();
buildModelFMU(Modelica.Blocks.Examples.PID_Controller); getErrorString();
```

and run it with omc from a terminal

```bash
omc test.mos &> test.log
```

You can try to activate the CMake FMUs and hope for the best. I don't have Mac available, so I can't help that much, but I can point you to the CMakeLists.txt which will (probably) throw the errors.

Run the following MOS-script to activate CMake FMUs and disable copying of runtime dependecies:

```modelica
loadModel(Modelica); getErrorString();
setCommandLineOptions("--fmuCMakeBuild=true --fmuRuntimeDepends=none"); getErrorString();
buildModelFMU(Modelica.Blocks.Examples.PID_Controller, platforms={"dynamic"}); getErrorString();
```

**Reply by arunraina2502** [reply]
As you suggested, here are the two logs. One without CMake and the second with CMake.

[test.log](https://github.com/OpenModelica/OpenModelica/files/10725843/test.log)
[test_cmake.log](https://github.com/OpenModelica/OpenModelica/files/10725847/test_cmake.log)

**Reply by AnHeuermann** [reply]
From the CMake build log:

> ld: unknown option: --no-undefined
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Apparently there is no linker option `--no-undefined` for clang available. You can try to remove it from the CMakeLists.txt.
It should be inside `Modelica.Blocks.Examples.PID_Controller.fmutmp/sources/CMakeLists.txt`.
Search for the part
```cmake
if(NOT ${CMAKE_VERSION} VERSION_LESS "3.13")
  target_link_options(${FMU_NAME} PRIVATE "LINKER:SHELL:--no-undefined")
endif()
```
and remove it.

Recompile the FMU. You need to provide the FMI headers, they are usually in your installation directory in 
`/path/to/OpenModelica/install_cmake/include/omc/FMI2`. Search for the file `fmi2_types.h`.

Run (change the <..> stuff):
```bash
cd Modelica.Blocks.Examples.PID_Controller.fmutmp/sources/
cmake -S. -B build_cmake -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DFMI_INTERFACE_HEADER_FILES_DIRECTORY=</path/to/fmi/headers>
cmake --build build_cmake --parallel <Nr. of cores> --target create_zip install
```

**Reply by arunraina2502** [reply]
Thanks for the response. I tried your steps exactly as given. It still throws errors.  [test2.log](https://github.com/OpenModelica/OpenModelica/files/10736293/test2.log)

**Reply by AnHeuermann** [reply]
Target `create_zip` is failing.

Maybe the issue is that you don't have any files in `Modelica.Blocks.Examples.PID_Controller.fmutmp/binaries`. In that case you need to run target `install` first or there is some problem building the dynamic library, but I can't see any error messages from that.

If you have the compiled dynmic library try playing around with the  `tar` command:

```CMake
add_custom_target(create_zip COMMAND
    ${CMAKE_COMMAND} -E tar "cfv" "../${CMAKE_PROJECT_NAME}.fmu" --format=zip
    "binaries/"
    "resources/"
    "sources/"
    "modelDescription.xml"
    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/../")
```
You can also just zip everything inside `Modelica.Blocks.Examples.PID_Controller.fmutmp` into `Modelica.Blocks.Examples.PID_Controller.fmu`.

**Reply by arunraina2502** [reply]
I was able to compile a dynamic library  inside `Modelica_Blocks_Examples_PID_Controller.fmutmp/binaries/darwin64/Modelica_Blocks_Examples_PID_Controller.dylib`. However, the fmu wasn't generated. I used zip command. 

I tried to execute this new FMU. This throws an error `Operator reinit may not be used in an algorithm section (use translation flag --allowNonStandardModelica=reinitInAlgorithms to ignore).` The modelica example itself works okay.

Besides, the method of compiling FMUs is cumbersome. The steps mentioned above are specific to Modelica example PID_Controller.

**Reply by AnHeuermann** [reply]
Can you tell me the zip command you used? And can you install tar, so the original CMakeLists will work?

> Besides, the method of compiling FMUs is cumbersome. The steps mentioned above are specific to Modelica example PID_Controller.

Well, if the changes to the CMake are working, you or myself can create a pull request to fix the issue on the main development branch.

**Reply by arunraina2502** [reply]
```
cd Modelica_Blocks_Examples_PID_Controller.fmutmp
zip -r Modelica_Blocks_Examples_PID_Controller.fmu .
```
I would like to combine all commands executed so far in a `CMakeList.txt` to make the compiling faster.  Will figure it out over the weekend. Is the error with executing FMU due to the zip command?

**Reply by AnHeuermann** [reply]
> I would like to combine all commands executed so far in a CMakeList.txt

If you like you can change the compiler itself to see if it works. I already made a change to exclude the linker flag on Apple, see the commit https://github.com/AnHeuermann/OpenModelica/commit/713ffb65e887caa765e9bdec6b6ca6c003fe2ff8 on my fork (not in the master yet).
You can change the CMakeList.txt inside the FMU by changing [OMCompiler/SimulationRuntime/fmi/export/buildproject/CMakeLists.txt.in](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/fmi/export/buildproject/CMakeLists.txt.in) and rebuilding OpenModelica (`cmake --build build_cmake --target install`) and then re-generating the FMU.

> Is the error with executing FMU due to the zip command?

I don't think so. The message comes from the broken legacy FMU import, see https://github.com/OpenModelica/OpenModelica/issues/9918. I hope the compilation itself is working.
You can try to use OMSimulator to verify that the FMU is running:

```bash
$ OMSimulator Modelica_Blocks_Examples_PID_Controller.fmu
```

**Reply by arunraina2502** [reply]
We can for the time being close this discussion. `OMEdit` cannot import and run FMUs. However, `OMSimulator` does the job. Thanks for the support.

The following steps you suggested solved the problem:

1. Open the file 
```
[OpenModelica/OMCompiler/SimulationRuntime/fmi/export/buildproject/CMakeLists.txt.in]
```
2. Comment in the following 
```
# if(NOT ${CMAKE_VERSION} VERSION_LESS "3.13")
#  target_link_options(${FMU_NAME} PRIVATE "LINKER:SHELL:--no-undefined")
# endif()
```
3.  Cleanup and re-install OpenModelica following instructions [here](https://github.com/OpenModelica/OpenModelica/blob/master/README.cmake.md#33-macos)
4. Create e.g. `build.mos` with the following commands
```
loadModel(Model_Name); getErrorString();
setCommandLineOptions("--fmuCMakeBuild=true --fmuRuntimeDepends=none"); getErrorString();
buildModelFMU(Model_Name, platforms={"dynamic"}); getErrorString();
```
5. Create an FMU by executing `omc build.mos` in terminal.
6. Run the newly created `Model_Name.fmu` with `OMSimulator` in terminal.
7. See the results of output file `model_res.mat` with `OMEdit` (or Scipy/Matlab/Octave).

**Comment by AnHeuermann** [comment]
With https://github.com/OpenModelica/OpenModelica/commit/a083b0eaad365f8c2398323d1a5d5a60c659203d this should work out of the box.
Make sure that you have `tar` installed and use the CMake FMU export:

```
loadModel(Modelica); getErrorString();
setCommandLineOptions("--fmuCMakeBuild=true"); getErrorString();
buildModelFMU(Modelica.Blocks.Examples.PID_Controller, platforms={"dynamic"}); getErrorString();
```
